### PR TITLE
lib: refactor status handling

### PR DESCRIPTION
**Pull request by liuzix**
- Extracted status handling logic to a separate package.
- Added the status content to the p2p message so that metastore is no longer read from each time the status is updated.

### CODE REVIEW

File: jobmaster/cvsJob/cvsJobMaster.go [hunk 1]
```diff
@@ -26,6 +26,8 @@ import (
 	"go.uber.org/atomic"
 )
 
+var _ lib.JobMasterImpl = &JobMaster{}
```
Comment (nongfushanquan) [inline jobmaster/cvsJob/cvsJobMaster.go:29]: why need this var ?

File: lib/base_jobmaster.go [hunk 2]
```diff
@@ -63,6 +63,7 @@ type JobMasterImpl interface {
 	OnWorkerOnline(worker WorkerHandle) error
 	OnWorkerOffline(worker WorkerHandle, reason error) error
 	OnWorkerMessage(worker WorkerHandle, topic p2p.Topic, message interface{}) error
+	OnWorkerStatusUpdated(worker WorkerHandle, newStatus *WorkerStatus) error
```
Comment (nongfushanquan) [inline lib/base_jobmaster.go:66]: is it necessary  to provide this method?

File: jobmaster/cvsJob/cvsJobMaster.go (hunk 1 above)
Comment (liuzix) [reply jobmaster/cvsJob/cvsJobMaster.go:30]: This is for testing purpose. We can move it somewhere else

File: lib/base_jobmaster.go (hunk 2 above)
Comment (liuzix) [reply lib/base_jobmaster.go]: The business logic might want to handle changes to the `ext` field. Providing this method seems the most convenient.

File: lib/statusutil/writer.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+			return
+		}
+		log.L().Warn("UpdateStatus failed",
+			zap.String("worker-id", w.workerID),
+			zap.String("master-id", w.masterInfo.MasterID()),
+			zap.String("master-node", w.masterInfo.MasterNode()),
+			zap.Int64("master-epoch", w.masterInfo.Epoch()))
+	}()
+
+	if reflect2.IsNil(w.lastStatus) || newStatus.HasChanged(w.lastStatus) {
```
Comment (amyangfei) [inline lib/statusutil/writer.go:61]: Where do we update `w.lastStatus`?

**Review by nongfushanquan — APPROVED** [review]

File: lib/worker.go [hunk 4]
```diff
@@ -307,8 +305,8 @@ func (w *DefaultBaseWorker) MetaKVClient() metaclient.KVClient {
 	return w.userMetaKVClient
 }
 
-func (w *DefaultBaseWorker) UpdateStatus(ctx context.Context, status WorkerStatus) error {
-	err := w.statusSender.AsyncSendStatus(ctx, status)
+func (w *DefaultBaseWorker) UpdateStatus(ctx context.Context, status libModel.WorkerStatus) error {
+	err := w.statusSender.UpdateStatus(ctx, &status)
```
Comment (amyangfei) [inline lib/worker.go:309]: Please add comment about the rate limit and delivery guarantee of `UpdateStatus`.
If worker sends messages with higher frequency than the processing rate than master, the overflow messages will be dropped in master side, is it right

**Review by amyangfei — APPROVED** [review]

**Pull request merged**